File: rag-backend/embed.js

require('dotenv').config({ path: require('path').resolve(__dirname, '../.env') });
const fs = require('fs');
const path = require('path');
const { QdrantClient } = require('@qdrant/js-client-rest');
const { GoogleGenerativeAI } = require('@google/generative-ai');

// Configuration
const COLLECTION_NAME = 'docusaurus-rag';
const VECTOR_SIZE = 768;
const DISTANCE = 'Cosine';
const DOCS_PATH = 'D:/4_quarter4_AIDD/Physical_AI_and_Humanoid_Robotics/docs';
const SUPPORTED_EXTENSIONS = ['.md', '.txt', '.pdf', '.html'];

// Initialize clients
let qdrantClient;
if (process.env.QDRANT_URL && process.env.QDRANT_API_KEY) {
  // Use cloud Qdrant instance
  qdrantClient = new QdrantClient({
    url: process.env.QDRANT_URL,
    apiKey: process.env.QDRANT_API_KEY,
    checkCompatibility: false
  });
  console.log("Using Qdrant Cloud instance");
} else {
  // Fallback to local Qdrant instance
  qdrantClient = new QdrantClient({
    url: 'http://localhost:6333',
    checkCompatibility: false
  });
  console.log("Using local Qdrant instance");
}

const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY || process.env.GEMINI_KEY);
const model = genAI.getGenerativeModel({ model: 'text-embedding-004' });

/**
 * Converts various file formats to plain text
 */
async function convertToText(filePath) {
  const ext = path.extname(filePath).toLowerCase();

  if (ext === '.txt') {
    return fs.readFileSync(filePath, 'utf8');
  } else if (ext === '.md') {
    const markdown = fs.readFileSync(filePath, 'utf8');
    // Simple markdown to text conversion (remove basic markdown syntax)
    return markdown
      .replace(/#{1,6}\s*/g, '') // Remove headers
      .replace(/\*\*(.*?)\*\*/g, '$1') // Remove bold
      .replace(/\*(.*?)\*/g, '$1') // Remove italic
      .replace(/\[(.*?)\]\(.*?\)/g, '$1') // Remove links
      .replace(/!\[.*?\]\(.*?\)/g, '') // Remove images
      .replace(/`{1,3}[^`]*`{1,3}/g, '') // Remove code blocks
      .replace(/^\s*[-*+]\s+/gm, '') // Remove list markers
      .replace(/^\s*\d+\.\s+/gm, '') // Remove numbered list markers
      .trim();
  } else if (ext === '.html') {
    const html = fs.readFileSync(filePath, 'utf8');
    // Simple HTML to text conversion (basic tag removal)
    return html
      .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '') // Remove script tags
      .replace(/<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>/gi, '') // Remove style tags
      .replace(/<[^>]+>/g, ' ') // Remove all other tags
      .replace(/&nbsp;/g, ' ') // Replace non-breaking spaces
      .replace(/&amp;/g, '&') // Replace ampersands
      .replace(/&lt;/g, '<') // Replace less than
      .replace(/&gt;/g, '>') // Replace greater than
      .replace(/\s+/g, ' ') // Normalize whitespace
      .trim();
  } else if (ext === '.pdf') {
    // For PDF, we need to install pdf-parse: npm install pdf-parse
    const pdfParse = require('pdf-parse');
    const pdfBuffer = fs.readFileSync(filePath);
    const pdfData = await pdfParse(pdfBuffer);
    return pdfData.text;
  } else {
    throw new Error(`Unsupported file extension: ${ext}`);
  }
}

/**
 * Splits text into chunks of 500-800 words
 */
function splitIntoChunks(text, minWords = 500, maxWords = 800) {
  const words = text.split(/\s+/);
  const chunks = [];
  let startIndex = 0;

  while (startIndex < words.length) {
    let endIndex = Math.min(startIndex + maxWords, words.length);

    // If we're near the end, make sure we don't go over maxWords
    if (endIndex - startIndex <= maxWords && endIndex === words.length) {
      chunks.push(words.slice(startIndex, endIndex).join(' '));
      break;
    }

    // If the chunk is too small, extend it
    if (endIndex - startIndex < minWords) {
      endIndex = Math.min(startIndex + minWords, words.length);
    }

    chunks.push(words.slice(startIndex, endIndex).join(' '));
    startIndex = endIndex;
  }

  return chunks;
}

/**
 * Generates embeddings using Gemini API
 */
// async function generateEmbedding(text) {
//   try {
//     const result = await model.embedContent(text);
//     return result.embedding.values; // Assuming this returns the vector array
//   } catch (error) {
//     console.error('Error generating embedding:', error);
//     throw error;
//   }
// }

async function generateEmbedding(text) {
  try {
    const result = await model.embedContent(text);
    console.log("Embedding length:", result.embedding.values.length); // 🔹 ye line add hui
    return result.embedding.values;
  } catch (error) {
    console.error('Error generating embedding:', error);
    throw error;
  }
}

/**
 * Checks if collection exists and creates it if it doesn't
 */
async function ensureCollectionExists() {
  try {
    await qdrantClient.getCollection(COLLECTION_NAME);
    console.log(`✅ Collection '${COLLECTION_NAME}' already exists`);
  } catch (error) {
    if (error.status === 404) {
      console.log(`📦 Creating collection '${COLLECTION_NAME}'...`);
      await qdrantClient.createCollection(COLLECTION_NAME, {
        vectors: {
          size: VECTOR_SIZE,
          distance: DISTANCE,
        },
      });
      console.log(`✅ Collection '${COLLECTION_NAME}' created successfully`);
    } else {
      throw error;
    }
  }
}

/**
 * Reads all files from the docs directory with supported extensions
 */
function getAllFiles(dirPath, arrayOfFiles = []) {
  const files = fs.readdirSync(dirPath);

  files.forEach((file) => {
    const filePath = path.join(dirPath, file);
    const stat = fs.statSync(filePath);

    if (stat.isDirectory()) {
      getAllFiles(filePath, arrayOfFiles);
    } else {
      const ext = path.extname(file).toLowerCase();
      if (SUPPORTED_EXTENSIONS.includes(ext)) {
        arrayOfFiles.push(filePath);
      }
    }
  });

  return arrayOfFiles;
}

/**
 * Main embedding function
 */
async function embedDocuments() {
  try {
    console.log('🚀 Starting document embedding process...');

    // Ensure collection exists
    await ensureCollectionExists();

    // Get all files to process
    const files = getAllFiles(DOCS_PATH);
    console.log(`📁 Found ${files.length} files to process`);

    if (files.length === 0) {
      console.log('⚠️ No supported files found in the docs directory');
      return;
    }

    let totalPointsInserted = 0;

    for (const file of files) {
      console.log(`\n📄 Processing file: ${file}`);

      try {
        // Convert file to text
        const text = await convertToText(file);

        if (!text.trim()) {
          console.log(`⚠️ Skipping empty file: ${file}`);
          continue;
        }

        // Split into chunks
        const chunks = splitIntoChunks(text);
        console.log(`✂️  Split into ${chunks.length} chunks`);

        // Process each chunk
        for (let i = 0; i < chunks.length; i++) {
          const chunk = chunks[i];

          if (!chunk.trim()) {
            continue; // Skip empty chunks
          }

          // Generate embedding
          const vector = await generateEmbedding(chunk);

          // Prepare point for insertion with numeric ID
          const point = {
            id: parseInt(`${Date.now()}${i.toString().padStart(3, '0')}`.slice(-16)), // Numeric ID based on timestamp
            vector: Array.from(vector), // Ensure vector is a proper array
            payload: {
              text: chunk.substring(0, 10000), // Limit text length to avoid payload size issues
              source: file,
              chunk_index: i,
            },
          };

          // Upsert the point with proper error handling
          try {
            await qdrantClient.upsert(COLLECTION_NAME, {
              points: [point],
              wait: true,  // Wait for operation to complete
            });
            console.log(`💾 Chunk ${i + 1}/${chunks.length} inserted`);
            totalPointsInserted++;
          } catch (upsertError) {
            console.error(`❌ Error inserting chunk ${i + 1} from ${file}:`, upsertError.message);
            // Print more detailed error info for debugging
            console.error(`   Error details:`, upsertError);
            // Continue to next chunk instead of stopping the entire process
            continue;
          }
        }

        console.log(`✅ Completed processing: ${file}`);
      } catch (error) {
        console.error(`❌ Error processing file ${file}:`, error.message);
        continue; // Continue with next file
      }
    }

    console.log(`\n📈 Summary: ${totalPointsInserted} points inserted into collection '${COLLECTION_NAME}'`);
    console.log('✅ Embedding Finished');
  } catch (error) {
    console.error('❌ Fatal error during embedding process:', error);
    process.exit(1);
  }
}

// Run the embedding process
if (require.main === module) {
  if (!process.env.GEMINI_API_KEY && !process.env.GEMINI_KEY) {
    console.error('❌ Error: GEMINI_API_KEY or GEMINI_KEY environment variable is required');
    console.error('Please set your Gemini API key: export GEMINI_API_KEY=your_api_key');
    process.exit(1);
  }

  embedDocuments();
}

module.exports = {
  convertToText,
  splitIntoChunks,
  generateEmbedding,
  ensureCollectionExists,
  getAllFiles,
  embedDocuments,
};

File: rag-backend/embed.py
"""
embed.py - Generate embeddings from the book content in docs folder
"""
import os
import logging
import dotenv
import google.generativeai as genai
from pathlib import Path
import re
import html
import PyPDF2
import time
from typing import List

# Load environment variables
dotenv.load_dotenv()

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Configure Gemini API
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY") or os.getenv("GEMINI_KEY")
if not GEMINI_API_KEY:
    raise ValueError("GEMINI_API_KEY or GEMINI_KEY environment variable is required")

genai.configure(api_key=GEMINI_API_KEY)

def convert_to_text(file_path: str) -> str:
    """
    Converts various file formats to plain text
    """
    file_path = Path(file_path)
    ext = file_path.suffix.lower()

    if ext == '.txt':
        with open(file_path, 'r', encoding='utf-8') as f:
            return f.read()
    elif ext == '.md':
        with open(file_path, 'r', encoding='utf-8') as f:
            markdown = f.read()
            # Simple markdown to text conversion (remove basic markdown syntax)
            text = re.sub(r'#{1,6}\s*', '', markdown)  # Remove headers
            text = re.sub(r'\*\*(.*?)\*\*', r'\1', text)  # Remove bold
            text = re.sub(r'\*(.*?)\*', r'\1', text)  # Remove italic
            text = re.sub(r'\[(.*?)\]\(.*?\)', r'\1', text)  # Remove links
            text = re.sub(r'!\[.*?\]\(.*?\)', '', text)  # Remove images
            text = re.sub(r'`{1,3}[^`]*`{1,3}', '', text)  # Remove code blocks
            text = re.sub(r'^\s*[-*+]\s+', '', text, flags=re.MULTILINE)  # Remove list markers
            text = re.sub(r'^\s*\d+\.\s+', '', text, flags=re.MULTILINE)  # Remove numbered list markers
            return text.strip()
    elif ext == '.html':
        with open(file_path, 'r', encoding='utf-8') as f:
            html_content = f.read()
            # Simple HTML to text conversion (basic tag removal)
            text = re.sub(r'<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>', '', html_content, flags=re.IGNORECASE)  # Remove script tags
            text = re.sub(r'<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>', '', text, flags=re.IGNORECASE)  # Remove style tags
            text = re.sub(r'<[^>]+>', ' ', text)  # Remove all other tags
            text = text.replace('&nbsp;', ' ')  # Replace non-breaking spaces
            text = html.unescape(text)  # Handle HTML entities
            text = re.sub(r'\s+', ' ', text)  # Normalize whitespace
            return text.strip()
    elif ext == '.pdf':
        with open(file_path, 'rb') as f:
            pdf_reader = PyPDF2.PdfReader(f)
            text = ""
            for page in pdf_reader.pages:
                text += page.extract_text() + "\n"
            return text
    else:
        raise ValueError(f"Unsupported file extension: {ext}")


def split_into_chunks(text: str, min_words: int = 500, max_words: int = 800) -> List[str]:
    """
    Splits text into chunks of 500-800 words
    """
    words = text.split()
    chunks = []
    start_index = 0

    while start_index < len(words):
        end_index = min(start_index + max_words, len(words))

        # If we're near the end, make sure we don't go over max_words
        if end_index - start_index <= max_words and end_index == len(words):
            chunks.append(' '.join(words[start_index:end_index]))
            break

        # If the chunk is too small, extend it
        if end_index - start_index < min_words:
            end_index = min(start_index + min_words, len(words))

        chunks.append(' '.join(words[start_index:end_index]))
        start_index = end_index

    return chunks


def generate_embedding(text: str) -> List[float]:
    """
    Generates embeddings using Gemini API
    """
    try:
        result = genai.embed_content(
            model="models/text-embedding-004",
            content=text,
        )
        logger.info(f"Embedding length: {len(result['embedding'])}")  # Debug line
        return result['embedding']
    except Exception as e:
        logger.error(f"Error generating embedding: {e}")
        raise e


def get_all_files(dir_path: str, supported_extensions: List[str]) -> List[str]:
    """
    Reads all files from the docs directory with supported extensions
    """
    dir_path = Path(dir_path)
    files = []

    for file_path in dir_path.rglob('*'):
        if file_path.is_file():
            ext = file_path.suffix.lower()
            if ext in supported_extensions:
                files.append(str(file_path))

    return files


def embed_documents(docs_path: str = "D:/4_quarter4_AIDD/Physical_AI_and_Humanoid_Robotics/docs"):
    """
    Main embedding function to convert docs to embeddings
    Returns list of embeddings with metadata
    """
    logger.info('🚀 Starting document embedding process...')

    SUPPORTED_EXTENSIONS = ['.md', '.txt', '.pdf', '.html']

    # Get all files to process
    files = get_all_files(docs_path, SUPPORTED_EXTENSIONS)
    logger.info(f'📁 Found {len(files)} files to process')

    if len(files) == 0:
        logger.info('⚠️ No supported files found in the docs directory')
        return []

    embeddings_data = []

    for file_path in files:
        logger.info(f'\n📄 Processing file: {file_path}')

        try:
            # Convert file to text
            text = convert_to_text(file_path)

            if not text.strip():
                logger.info(f'⚠️ Skipping empty file: {file_path}')
                continue

            # Split into chunks
            chunks = split_into_chunks(text)
            logger.info(f'✂️  Split into {len(chunks)} chunks')

            # Process each chunk
            for i, chunk in enumerate(chunks):
                if not chunk.strip():
                    continue  # Skip empty chunks

                # Generate embedding
                vector = generate_embedding(chunk)

                # Add to results
                embeddings_data.append({
                    'embedding': vector,
                    'text': chunk[:10000],  # Limit text length to avoid payload size issues
                    'source': file_path,
                    'chunk_index': i
                })

                logger.info(f'✅ Chunk {i + 1}/{len(chunks)} processed')

            logger.info(f'✅ Completed processing: {file_path}')
        except Exception as error:
            logger.error(f'❌ Error processing file {file_path}: {error}')
            continue  # Continue with next file

    logger.info(f'\n📈 Summary: {len(embeddings_data)} embeddings generated')
    logger.info('✅ Embedding Generation Finished')
    return embeddings_data


if __name__ == "__main__":
    # Generate embeddings for all documents
    embeddings = embed_documents()
    print(f"Generated {len(embeddings)} embeddings from the book content")